### PR TITLE
[NOCOMMIT] SIMD repcode

### PR DESCRIPTION
**Pull request by terrelln**
Creating & closing a PR so the branch doesn't get lost.

Before:
```
 5#silesia.tar       : 211957760 ->  63995184 (3.312), 107.7 MB/s , 962.1 MB/s 
 6#silesia.tar       : 211957760 ->  62897060 (3.370),  84.2 MB/s , 975.2 MB/s 
 7#silesia.tar       : 211957760 ->  61339837 (3.455),  60.8 MB/s ,1047.2 MB/s 
 8#silesia.tar       : 211957760 ->  60759491 (3.488),  48.5 MB/s ,1076.8 MB/s

 5#xml               :   5345280 ->    613377 (8.715), 212.5 MB/s ,1687.1 MB/s 
 6#xml               :   5345280 ->    581003 (9.200), 177.2 MB/s ,1795.9 MB/s 
 7#xml               :   5345280 ->    552163 (9.681), 125.1 MB/s ,2012.0 MB/s 
 8#xml               :   5345280 ->    536286 (9.967), 100.2 MB/s ,2084.4 MB/s 

 5#nci               :  33553445 ->   2725030 (12.31), 294.0 MB/s ,1709.9 MB/s 
 6#nci               :  33553445 ->   2622009 (12.80), 251.8 MB/s ,1788.7 MB/s 
 7#nci               :  33553445 ->   2461144 (13.63), 156.5 MB/s ,2138.8 MB/s 
 8#nci               :  33553445 ->   2376915 (14.12), 122.9 MB/s ,2210.2 MB/s 

 5#kennedy.xls       :   1029744 ->    111783 (9.212), 206.9 MB/s ,1060.8 MB/s 
 6#kennedy.xls       :   1029744 ->    111218 (9.259), 190.7 MB/s ,1066.8 MB/s 
 7#kennedy.xls       :   1029744 ->    108051 (9.530), 109.6 MB/s ,1080.3 MB/s 
 8#kennedy.xls       :   1029744 ->    111092 (9.269), 102.6 MB/s ,1055.7 MB/s 
```

After:
```
 5#silesia.tar       : 211957760 ->  63749370 (3.325), 105.8 MB/s , 930.2 MB/s 
 6#silesia.tar       : 211957760 ->  62685213 (3.381),  83.8 MB/s , 944.2 MB/s 
 7#silesia.tar       : 211957760 ->  61145057 (3.466),  60.9 MB/s ,1016.8 MB/s 
 8#silesia.tar       : 211957760 ->  60597877 (3.498),  47.8 MB/s ,1046.0 MB/s

 5#xml               :   5345280 ->    602707 (8.869), 211.0 MB/s ,1666.9 MB/s 
 6#xml               :   5345280 ->    572218 (9.341), 178.7 MB/s ,1764.7 MB/s 
 7#xml               :   5345280 ->    543176 (9.841), 127.4 MB/s ,1967.8 MB/s 
 8#xml               :   5345280 ->    526805 (10.15), 101.7 MB/s ,2050.4 MB/s 

 5#nci               :  33553445 ->   2664702 (12.59), 279.0 MB/s ,1568.5 MB/s 
 6#nci               :  33553445 ->   2573628 (13.04), 242.2 MB/s ,1625.6 MB/s 
 7#nci               :  33553445 ->   2424801 (13.84), 152.2 MB/s ,1953.1 MB/s 
 8#nci               :  33553445 ->   2341188 (14.33), 119.9 MB/s ,2022.5 MB/s 

 5#kennedy.xls       :   1029744 ->    109838 (9.375), 179.1 MB/s , 990.1 MB/s 
 6#kennedy.xls       :   1029744 ->    109816 (9.377), 169.5 MB/s , 995.5 MB/s 
 7#kennedy.xls       :   1029744 ->    106771 (9.644), 103.9 MB/s ,1081.7 MB/s 
 8#kennedy.xls       :   1029744 ->    110285 (9.337),  93.7 MB/s ,1052.9 MB/s 
```

Note the loss of decompression speed. Compression speed drops when the repcode density is high.

### CODE REVIEW

**Comment by Cyan4973** [comment]
`xml` compression gains are fairly impressive.

I presume increasing repcode density, on top of improving compression ratio, also results in more and smaller matches, explaining the speed impact.

One possibility to mitigate this would be to only accept repcode of a larger minimum length (typically in line with the `mml` parameter), but it's unclear if that fits well with SIMD code.

**Comment by terrelln** [comment]
> I presume increasing repcode density, on top of improving compression ratio, also results in more and smaller matches, explaining the speed impact.

Yeah, smaller matches, and more repcodes, which decrease the predictability of the `is repcode` branch in the decoder. And also decreases the predictability of the `is repcode 0` branch.

**Comment by terrelln** [comment]
> One possibility to mitigate this would be to only accept repcode of a larger minimum (typically in line with the mml parameter), but it's unclear if that fits well with SIMD code.

It could be done with AVX2 easily. Since you could load up to 8-bytes per repcode. I'm currently using gather from AVX2, but that can be emulated without AVX without too much of a speed loss.

At this point I'm going to leave this as a possible optimization for the future, especially when we have the new lazy match finder from @senhuang42. We may decide it is worth sacrificing a little of the speed we gained to have stronger repcode search.

I also am going to test it on some Managed Compression data, because we know we have a higher repcode density there.

File: lib/compress/zstd_lazy.c [hunk 1]
```diff
@@ -973,37 +1141,55 @@ ZSTD_compressBlock_lazy_generic(
...
-                    const BYTE* repMatchEnd = repIndex < prefixLowestIndex ? dictEnd : iend;
-                    size_t const mlRep = ZSTD_count_2segments(ip+4, repMatch+4, iend, repMatchEnd, prefixLowest) + 4;
+            {
+                ZS_Vec128 const rep32 = ZS_Repcode_gather32(reps, ip);
+                BYTE const* const rip = ip;
+                U32 const repcode = ZS_Repcode_search(reps, rep32, rip);
+                if (repcode != 0) {
+                    BYTE const* repMatch = rip + ZS_Repcode_get(reps, repcode);
+                    size_t const mlRep = ZSTD_count(rip+4, repMatch + 4, iend) + 4;
                     int const gain2 = (int)(mlRep * 3);
```
Comment (Cyan4973) [inline lib/compress/zstd_lazy.c:1163]: minor FYI : 
this formula, which makes a crude selection of current `repcode` versus previous "full offset" match based on a fixed cost evaluation, is expecting `mlCode == 0`. 
Now, if `mlCode` can also be `1` or `2`, this formula might need an adjustment.

Note that it's merely a small optimization, no big change is expected.

**Pull request closed without merging**